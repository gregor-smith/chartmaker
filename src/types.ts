import {
    Record as Record_,
    Literal,
    String,
    Union,
    InstanceOf,
    Boolean,
    Static,
    Number as Number_,
    Array as Array_,
    Runtype,
    Partial as Partial_
} from 'runtypes'

import {
    CHART_ALBUMS_COUNT,
    MAX_SCREENSHOT_SCALE,
    MAX_COLLAGE_ROWS_X,
    MAX_COLLAGE_ROWS_Y,
    STATE_VERSION,
    EXPORT_CHART_PLACEHOLDER
} from '@/constants'


const Integer = Number_.withConstraint(Number.isSafeInteger)


function IntegerRange(minimum: number, maximum = Number.MAX_SAFE_INTEGER) {
    return Integer.withConstraint(number => number >= minimum && number <= maximum)
}


function FixedSizeArray<T extends Runtype>(element: T, size: number) {
    return Array_(element).withConstraint(array => array.length === size)
}


function NonEmptyArray<T extends Runtype>(element: T) {
    return Array_(element).withConstraint(array => array.length > 0)
}


const NonEmptyString = String.withConstraint(string => string.length !== 0)


const PositiveInteger = IntegerRange(0)


const V1PlaceholderAlbum = Record_({
    placeholder: Literal(true),
    id: PositiveInteger
})


const SearchAlbum = Record_({
    name: NonEmptyString,
    url: NonEmptyString
})


const V1NamedAlbum = SearchAlbum.And(
    Record_({
        placeholder: Literal(false),
        id: PositiveInteger
    })
)
const V2NamedAlbum = SearchAlbum.And(
    Record_({
        id: PositiveInteger
    })
)
export const NamedAlbum = V2NamedAlbum


const V1Album = V1NamedAlbum.Or(V1PlaceholderAlbum)
const V2Album = V2NamedAlbum.Or(PositiveInteger)
const Album = V2Album


const ChartShape = Union(
    Record_({
        tag: Literal('Top'),
        size: Union(
            Literal(40),
            Literal(42),
            Literal(100)
        )
    }),
    Record_({
        tag: Literal('Collage')
    })
)


const V1Chart = Record_({
    name: NonEmptyString,
    albums: FixedSizeArray(V1Album, CHART_ALBUMS_COUNT),
    shape: ChartShape,
    rowsX: IntegerRange(1, MAX_COLLAGE_ROWS_X),
    rowsY: IntegerRange(1, MAX_COLLAGE_ROWS_Y)
})
const V2Chart = Record_({
    ...V1Chart.fields,
    albums: FixedSizeArray(Album, CHART_ALBUMS_COUNT)
})
const Chart = V2Chart


const SearchState = Union(
    Record_({
        tag: Literal('Waiting')
    }),
    Record_({
        tag: Literal('Loading'),
        controller: InstanceOf(AbortController)
    }),
    Record_({
        tag: Literal('Complete'),
        albums: NonEmptyArray(SearchAlbum)
    }),
    Record_({
        tag: Literal('Error'),
        message: String
    })
).And(
    Record_({
        query: String
    })
)


const ScreenshotState = Record_({
    loading: Boolean,
    scale: IntegerRange(1, MAX_SCREENSHOT_SCALE)
})


const ExportChartAlbum = SearchAlbum.Or(Literal(EXPORT_CHART_PLACEHOLDER))

const ExportChartShape = ChartShape.alternatives[0].Or(
    ChartShape.alternatives[1].And(
        Record_({
            rowsX: V2Chart.fields.rowsX,
            rowsY: V2Chart.fields.rowsY,
        })
    )
)

const ExportChart = Record_({
    name: V2Chart.fields.name,
    albums: FixedSizeArray(ExportChartAlbum, CHART_ALBUMS_COUNT),
    shape: ExportChartShape
})


export const V1State = Record_({
    apiKey: String,
    charts: NonEmptyArray(V1Chart),
    activeChartIndex: PositiveInteger,
    search: SearchState,
    screenshot: ScreenshotState
}).And(
    Partial_({
        version: Literal(1),
        highlightedID: PositiveInteger
    })
)
export const V2State = Record_({
    ...V1State.intersectees[0].fields,
    version: Literal(2),
    charts: NonEmptyArray(V2Chart)
}).And(
    Partial_({
        highlightedID: PositiveInteger
    })
)
const V3State = Record_({
    ...V2State.intersectees[0].fields,
    version: Literal(STATE_VERSION)
}).And(
    Partial_({
        ...V2State.intersectees[1].fields,
        viewing: ExportChart
    })
)
export const State = V3State


export type NamedAlbum = Static<typeof NamedAlbum>
export type SearchAlbum = Static<typeof SearchAlbum>
export type Album = Static<typeof Album>
export type ChartShape = Static<typeof ChartShape>
export type Chart = Static<typeof Chart>
export type SearchState = Static<typeof SearchState>
export type ScreenshotState = Static<typeof ScreenshotState>
export type State = Static<typeof State>
export type V1State = Static<typeof V1State>
export type V2State = Static<typeof V2State>
export type ExportChart = Static<typeof ExportChart>
export type ExportChartShape = Static<typeof ExportChartShape>
export type ExportChartAlbum = Static<typeof ExportChartAlbum>
