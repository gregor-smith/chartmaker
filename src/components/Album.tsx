import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
