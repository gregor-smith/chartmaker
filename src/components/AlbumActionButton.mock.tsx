const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
