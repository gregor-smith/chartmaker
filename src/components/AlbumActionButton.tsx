import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
