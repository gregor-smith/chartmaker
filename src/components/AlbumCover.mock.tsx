import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
