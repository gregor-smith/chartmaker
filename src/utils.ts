import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
