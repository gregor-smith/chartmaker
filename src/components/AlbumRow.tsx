import React, { FC } from 'react'
import { css } from 'emotion'

import { Album as AlbumDetails } from '../reducer'
import { Album } from './Album'


type Props = {
    albums: (AlbumDetails | null)[]
    size: number
    from?: number
    to?: number
}


const style = css({
    display: 'flex'
})


export const AlbumRow: FC<Props> = ({ albums, from = 0, to = albums.length, size }) => {
    const row: JSX.Element[] = []
    for (let index = from; index < to; index++) {
        const album = albums[index] ?? null
        row.push(<Album key={index} details={album} size={size}/>)
    }
    return (
        <div className={style}>
            {row}
        </div>
    )
}
