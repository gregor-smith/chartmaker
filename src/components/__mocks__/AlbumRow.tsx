import React, { FC } from 'react'

import { AlbumRowProps } from '@/components/AlbumRow'


export const AlbumRow: FC<AlbumRowProps> = ({ albums, size }) => {
    const json = JSON.stringify(albums)
    return (
        <div className='test-album-row'>
            {`Albums: ${json}`}
            {`Size: ${size}`}
        </div>
    )
}
