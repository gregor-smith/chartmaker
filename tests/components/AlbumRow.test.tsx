import React from 'react'
import { render } from 'react-dom'

import { Album, Size } from '@/types'
import { AlbumRow } from '@/components/AlbumRow'

import { RenderContainer, ignore } from '../utils'


jest.mock('@/components/ChartAlbumCover')


const container = new RenderContainer()


test('renders album covers', () => {
    const albums: Album[] = [...Array(5).keys()].map(index => ({
        placeholder: true,
        id: index
    }))

    render(
        <AlbumRow dispatch={ignore}
            albums={albums}
            size={'5rem' as Size}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})
