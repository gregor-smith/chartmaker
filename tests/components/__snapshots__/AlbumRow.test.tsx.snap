// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders album covers 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-0"
  >
    <div
      class="test-chart-album-cover"
    >
      Album: {"placeholder":true,"id":0}
      Size: 5rem
    </div>
    <div
      class="test-chart-album-cover"
    >
      Album: {"placeholder":true,"id":1}
      Size: 5rem
    </div>
    <div
      class="test-chart-album-cover"
    >
      Album: {"placeholder":true,"id":2}
      Size: 5rem
    </div>
    <div
      class="test-chart-album-cover"
    >
      Album: {"placeholder":true,"id":3}
      Size: 5rem
    </div>
    <div
      class="test-chart-album-cover"
    >
      Album: {"placeholder":true,"id":4}
      Size: 5rem
    </div>
  </div>
</div>
`;
